asyncpg/connect_utils.py

```diff
@@ -678,20 +738,58 @@ def data_received(self, data):
                     'rejected SSL upgrade'.format(
                         host=self.host, port=self.port)))
 
-    def connection_lost(self, exc):
+    def connection_lost(self, exc: typing.Optional[Exception]) -> None:
         if not self.on_data.done():
             if exc is None:
                 exc = ConnectionError('unexpected connection_lost() call')
             self.on_data.set_exception(exc)
 
 
-async def _create_ssl_connection(protocol_factory, host, port, *,
-                                 loop, ssl_context, ssl_is_advisory=False):
-
-    tr, pr = await loop.create_connection(
-        lambda: TLSUpgradeProto(loop, host, port,
-                                ssl_context, ssl_is_advisory),
-        host, port)
+@typing.overload
+async def _create_ssl_connection(
+    protocol_factory: typing.Callable[[], _Protocol],
+    host: str,
+    port: int,
+    *,
+    loop: asyncio.AbstractEventLoop,
+    ssl_context: ssl_module.SSLContext,
+    ssl_is_advisory: typing.Optional[bool] = False
+) -> _TPTupleType[_Protocol]:
+    ...
+
+
+@typing.overload
+async def _create_ssl_connection(
+    protocol_factory: typing.Callable[[], '_CancelProto'],
+    host: str,
+    port: int,
+    *,
+    loop: asyncio.AbstractEventLoop,
+    ssl_context: ssl_module.SSLContext,
+    ssl_is_advisory: typing.Optional[bool] = False
+) -> _TPTupleType['_CancelProto']:
+    ...
+
+
+async def _create_ssl_connection(
+    protocol_factory: typing.Union[
+        typing.Callable[[], _Protocol],
+        typing.Callable[[], '_CancelProto']
+    ],
+    host: str,
+    port: int,
+    *,
+    loop: asyncio.AbstractEventLoop,
+    ssl_context: ssl_module.SSLContext,
+    ssl_is_advisory: typing.Optional[bool] = False
+) -> _TPTupleType[typing.Union[_Protocol, '_CancelProto']]:
```

Comment:
Pylance gives here following warning:
```
Could not specialize type "_TPTupleType[_Protocol@_TPTupleType]"
  Type "_Protocol@_create_ssl_connection | _CancelProto" cannot be assigned to type "Protocol"
```

If mypy has no issues with this - maybe this is a Pylance problem

Reply:
Same as above. I'm investigating.

Reply:
I was able to fix this with a few small changes I pushed